x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DatePart.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.sql.expression.function.scalar.datetime;
+
+import org.elasticsearch.xpack.sql.expression.Expression;
+import org.elasticsearch.xpack.sql.expression.Nullability;
+import org.elasticsearch.xpack.sql.expression.function.scalar.BinaryScalarFunction;
+import org.elasticsearch.xpack.sql.expression.function.scalar.datetime.DateTimeProcessor.DateTimeExtractor;
+import org.elasticsearch.xpack.sql.tree.NodeInfo;
+import org.elasticsearch.xpack.sql.tree.Source;
+import org.elasticsearch.xpack.sql.type.DataType;
+
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.temporal.ChronoField;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.sql.expression.function.scalar.datetime.NonIsoDateTimeProcessor.NonIsoDateTimeExtractor;
+
+public class DatePart extends BinaryDateTimeFunction {
+
+    public enum Part implements DateTimeField {
+        YEAR(DateTimeExtractor.YEAR::extract, "years", "yyyy", "yy"),
+        QUARTER(QuarterProcessor::quarter, "quarters", "qq", "q"),
+        MONTH(DateTimeExtractor.MONTH_OF_YEAR::extract, "months", "mm", "m"),
+        DAYOFYEAR(DateTimeExtractor.DAY_OF_YEAR::extract, "dy", "y"),
+        DAY(DateTimeExtractor.DAY_OF_MONTH::extract, "days", "dd", "d"),
+        WEEK(NonIsoDateTimeExtractor.WEEK_OF_YEAR::extract, "weeks", "wk", "ww"),
+        WEEKDAY(NonIsoDateTimeExtractor.DAY_OF_WEEK::extract, "weekdays", "dw"),
+        HOUR(DateTimeExtractor.HOUR_OF_DAY::extract, "hours", "hh"),
+        MINUTE(DateTimeExtractor.MINUTE_OF_HOUR::extract, "minutes", "mi", "n"),
+        SECOND(DateTimeExtractor.SECOND_OF_MINUTE::extract, "seconds", "ss", "s"),
+        MILLISECOND(dt -> dt.get(ChronoField.MILLI_OF_SECOND), "milliseconds", "ms"),
+        MICROSECOND(dt -> dt.get(ChronoField.MICRO_OF_SECOND), "microseconds", "mcs"),
+        NANOSECOND(ZonedDateTime::getNano, "nanoseconds", "ns"),
+        TZOFFSET(dt -> dt.getOffset().getTotalSeconds() / 60, "tz");
+
+        private static final Map<String, Part> NAME_TO_PART;
+        private static final List<String> VALID_VALUES;
+
+        static {
+            NAME_TO_PART = DateTimeField.initializeResolutionMap(values());
+            VALID_VALUES = DateTimeField.initializeValidValues(values());
+        }
+
+        private Set<String> aliases;
```

Comment:
I would change the definition order to be extractFunction, aliases to match the constructor parameters order and initialization.